linera-core/src/client.rs

```diff
@@ -1737,10 +1864,13 @@ where
         }
         let manager = *info.manager;
         // Drop the pending block if it is outdated.
-        if let Some(block) = &self.pending_block {
-            if block.height != info.next_block_height {
-                self.clear_pending_block();
-            }
+        // Caveat editor: this cannot be an `if let` as the borrow of `state()` will be
+        // extended over the body, causing a deadlock.
+        if matches!(
+            self.state().pending_block,
+            Some(ref block) if block.height != info.next_block_height,
+        ) {
+            self.clear_pending_block();
```

Comment:
Please let's stick with the current style in the codebase: `matches!(&self.state().pending_block, Some(block) ...)`

Perhaps a more self-explaining way to write this is:
```
{
    let state = self.state();
    if let Some(block) = &state.pending_block { .. }
}
```

Reply:
Self-explanatory — but will cause a deadlock!  The code here has been very carefully written, to the point that I had to add a warning comment.  I think that relying on the match-ergonomics syntax sugar here is probably safe, though — I'll give it a go and see if the tests like it.

Reply:
Perhaps a variant of this like

```rust
{
    let state = self.state();
    if let Some(block) = &state.pending_block {
        if block.height != info.next_block_height {
            drop(block);
            drop(state);
            self.clear_pending_block();
        }
    }
}
```

but I know we prefer to avoid additional levels of indentation where possible.  We can try this if you prefer it though.

Reply:
I like the `drop(state)` for sure. More explicit and kinda removes the need for a comment as a result.

Reply:
I think I'll leave the comment there just in case, though.